janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphMixedIndexCountStrategy.java

```diff
@@ -0,0 +1,147 @@
+import org.apache.tinkerpop.gremlin.process.traversal.step.map.CountGlobalStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.map.GraphStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.map.NoOpBarrierStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.sideEffect.IdentityStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.util.HasContainer;
+import org.apache.tinkerpop.gremlin.process.traversal.strategy.AbstractTraversalStrategy;
+import org.apache.tinkerpop.gremlin.process.traversal.util.TraversalHelper;
+import org.apache.tinkerpop.gremlin.structure.T;
+import org.janusgraph.core.JanusGraphTransaction;
+import org.janusgraph.core.MixedIndexCountQuery;
+import org.janusgraph.graphdb.query.JanusGraphPredicateUtils;
+
+import java.util.Collections;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.janusgraph.graphdb.tinkerpop.optimize.HasStepFolder.validFoldInHasContainer;
+
+/**
+ * If the query can be satisfied by a single mixed index query, and the query is followed by a count step, then
+ * this strategy replaces original step with {@link JanusGraphMixedIndexCountStep}, which fires a count query against
+ * mixed index backend without retrieving all elements
+ *
+ * @author Boxuan Li (liboxuan@connect.hku.hk)
+ */
+public class JanusGraphMixedIndexCountStrategy extends AbstractTraversalStrategy<TraversalStrategy.ProviderOptimizationStrategy>
+    implements TraversalStrategy.ProviderOptimizationStrategy {
+    private static final JanusGraphMixedIndexCountStrategy INSTANCE = new JanusGraphMixedIndexCountStrategy();
+
+    private JanusGraphMixedIndexCountStrategy() {
+    }
+
+    @Override
+    public void apply(final Traversal.Admin<?, ?> traversal) {
+        if (TraversalHelper.onGraphComputer(traversal))
+            return;
+
+        TraversalHelper.getStepsOfClass(GraphStep.class, traversal).forEach(originalGraphStep -> {
+            if (originalGraphStep.getIds() == null || originalGraphStep.getIds().length == 0) {
+                buildMixedIndexCountStep(originalGraphStep, traversal);
+            }
+        });
+    }
+
+    private void buildMixedIndexCountStep(final GraphStep originalGraphStep, final Traversal.Admin<?, ?> traversal) {
+        if (!hasCountGlobalStep(originalGraphStep)) {
+            return;
+        }
+        final JanusGraphMixedIndexCountStep<?> directQueryCountStep = new JanusGraphMixedIndexCountStep<>(traversal);
+        // we shouldn't really fold in, but only check it, because we don't want to break the original traversal,
+        // as this optimizer might fail
+        if (!foldInHasContainer(originalGraphStep, directQueryCountStep, traversal, traversal)) {
+            return;
+        }
+        final JanusGraphTransaction tx = JanusGraphTraversalUtil.getTx(traversal);
+        MixedIndexCountQuery mixedIndexCountQuery = directQueryCountStep.buildGlobalGraphIndexQuery(originalGraphStep, tx);
```

Comment:
It took me about half an hour to figure out that _this_ is actually the line that checks if the optimization should be applied at all. I tried to find the line that clearly differentiates between mixed and composite indexes. The fact that this is buried deeply inside the index selection part is a bit of an obfuscation of what actually happens here.
Now that it became clear to me, it would be good to at least point out exactly what happens here.